devfile/devworkspace-operator
# Add new field for DWOC for DevWorkspace pod's runtimeClass

### What does this PR do?
Introduces a new field for the DWOC for the DevWorkspace pod's runtimeClass

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/1286

### Is it tested? How?

To test this PR,

1. Create the plain-devworkspace DevWorkspace on the cluster
```
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: plain-devworkspace
spec:
  started: true
  routingClass: 'basic'
  template:
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
```

2. Verify that the workspace pod starts and runs properly.

3. Create this DWOC:
```
apiVersion: controller.devfile.io/v1alpha1
config:
  workspace:
    runtimeClassName: test
kind: DevWorkspaceOperatorConfig
metadata:
  name: devworkspace-operator-config
  namespace: openshift-operators
```

4. Restart the plain-devworkspace DevWorkspace.
The DevWorkspace startup should fail with:
```
'Error creating DevWorkspace deployment: Detected unrecoverable deployment condition: FailedCreate pods "workspace9b9d2d153f3747de-55bb88dd5b-" is forbidden: pod rejected: RuntimeClass "test" not found'
```

5. To verify that the runtimeClassName attribute takes precedence over the DWOC-defined runtimeClassName, create the following DevWorkspace:
```
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: plain-devworkspace-2
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/runtime-class: dw-class
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
```

6. The DevWorkspace startup should fail with:
```
'Error creating DevWorkspace deployment: Detected unrecoverable deployment condition: FailedCreate pods "workspaceb61b62899c1549aa-55bcc6dbfc-" is forbidden: pod rejected: RuntimeClass "dw-class" not found'
```

### PR Checklist
